# Name : Krishnakant Singhal
## Data Science and Business Analytics Intern (The Sparks Foundation)

### Task : Prediction using Decision Tree Algorithm (Level - Intermediate)

### Imporing Essential Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Loading Dataset using Pandas read_csv Function

In [2]:
df = pd.read_csv("C:\\Users\\krishna\\machine learning projects\\Internship TSF\\Task 2\\Iris.csv")
df.head()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

### Checking wheather there is a nan value in dataset

In [3]:
df.isnull().sum()

Id               0
SepalLengthCm    0
SepalWidthCm     0
PetalLengthCm    0
PetalWidthCm     0
Species          0
dtype: int64

### Checking the datatype of values in each row of dataframe

In [4]:
df.dtypes

Id                 int64
SepalLengthCm    float64
SepalWidthCm     float64
PetalLengthCm    float64
PetalWidthCm     float64
Species           object
dtype: object

## Spliting Data into features and Target Values

In [5]:
X = df.drop(['Id', 'Species'], axis = 1)
y = df['Species']

## Spliting Data into Training and Testing using sklearn train_test_split module

In [6]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.1, random_state = 53)

print("The shape of X_train: ", X_train.shape)
print("The shape of X_test: ", X_test.shape)
print("The shape of y_train: ", y_train.shape)
print("The shape of y_test: ", y_test.shape)

The shape of X_train:  (135, 4)
The shape of X_test:  (15, 4)
The shape of y_train:  (135,)
The shape of y_test:  (15,)


### Checking the type of the splitted data

In [7]:
type(X_train), type(X_test), type(y_train), type(y_test)

(pandas.core.frame.DataFrame,
 pandas.core.frame.DataFrame,
 pandas.core.series.Series,
 pandas.core.series.Series)

## Feature Scaling data using sklearn StandardScaler module

In [8]:
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
sc.fit(X_train, X_test)
X_train_sc = sc.transform(X_train)
X_test_sc = sc.transform(X_test)

In [9]:
X_train_sc, X_test_sc

(array([[-0.40959573, -1.54300365, -0.01743847, -0.24713985],
        [ 0.47010419, -2.00385651,  0.44191629,  0.41222536],
        [ 0.47010419, -0.62129793,  0.61417433,  0.80784449],
        [ 1.3498041 ,  0.30040779,  1.13094844,  1.46720971],
        [-1.03795281,  0.53083422, -1.33808341, -1.3021242 ],
        [ 0.09308994,  0.30040779,  0.61417433,  0.80784449],
        [ 1.09846127,  0.53083422,  1.13094844,  1.20346362],
        [-0.28392431, -0.3908715 , -0.07485781,  0.14847928],
        [-0.78660998,  0.76126065, -1.33808341, -1.3021242 ],
        [-0.03258148, -0.85172436,  0.09740022,  0.01660624],
        [-0.91228139,  0.99168708, -1.33808341, -1.3021242 ],
        [-1.16362422,  0.06998136, -1.28066406, -1.43399724],
        [ 0.34443277, -0.62129793,  0.15481957,  0.14847928],
        [-0.28392431, -0.62129793,  0.67159368,  1.07159058],
        [-1.54063847,  0.06998136, -1.28066406, -1.3021242 ],
        [ 0.5957756 ,  0.53083422,  1.30320647,  1.7309558 ],
        

## Applyig Decision Tree Classifier for pridiction with 'gini' criterion on non-feature scaled data

In [10]:
from sklearn.tree import DecisionTreeClassifier

dec_class_gini = DecisionTreeClassifier(criterion = 'gini')
dec_class_gini.fit(X_train, y_train)
dec_class_gini.score(X_test, y_test)

0.9333333333333333

## Applyig Decision Tree Classifier for pridiction with 'entropy' criterion on non feature scaled data

In [11]:
dec_class_en = DecisionTreeClassifier(criterion = 'entropy')
dec_class_en.fit(X_train, y_train)
dec_class_en.score(X_test, y_test)

0.9333333333333333

## Applyig Decision Tree Classifier for pridiction with 'gini' criterion on feature scaled data

In [12]:
dec_class_sc_gini = DecisionTreeClassifier(criterion = 'gini')
dec_class_sc_gini.fit(X_train_sc, y_train)
dec_class_sc_gini.score(X_test, y_test)

0.3333333333333333

## Applyig Decision Tree Classifier for pridiction with 'entropy' criterion on feature scaled data

In [13]:
dec_class_sc_en = DecisionTreeClassifier(criterion = 'entropy')
dec_class_sc_en.fit(X_train_sc, y_train)
dec_class_sc_en.score(X_test, y_test)

0.3333333333333333

## We will take model of Classifier without feature scaled data and with entropy criterion because this is giving higher accuracy than other (93.33 %)

## Predicting the value of Proper Class for Testing data

In [14]:
dec_class_en.predict(X_test)

array(['Iris-setosa', 'Iris-setosa', 'Iris-virginica', 'Iris-versicolor',
       'Iris-versicolor', 'Iris-virginica', 'Iris-virginica',
       'Iris-setosa', 'Iris-versicolor', 'Iris-virginica', 'Iris-setosa',
       'Iris-versicolor', 'Iris-virginica', 'Iris-setosa',
       'Iris-virginica'], dtype=object)